src/main/java/fi/aalto/cs/intellij/common/Resources.java

```diff
@@ -0,0 +1,71 @@
+package fi.aalto.cs.intellij.common;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A class that provides methods to read the resource files of the plugin application.
+ * Normally, this class should be accessed via {@code Resources.SingletonUtility.INSTANCE}.
+ */
+class Resources {
+
+  public static final Logger LOGGER = LoggerFactory.getLogger(Resources.class);
+
+  private ResourceProvider resourceProvider;
+
+  Resources(@NotNull ClassLoader classLoader) {
+    this(classLoader::getResourceAsStream);
+  }
+
+  Resources(@NotNull ResourceProvider resourceProvider) {
+    this.resourceProvider = resourceProvider;
+  }
+
+  /**
+   * Returns properties from a properties resource file.
+   * @param resourceName Name of a resource.
+   * @return Contents of the resource in a {@link Properties} object.
+   */
+  @Nullable
+  public Properties getProperties(@NotNull String resourceName) {
+    try (InputStream stream = resourceProvider.getResourceAsStream(resourceName)) {
+      return getProperties(stream);
+    } catch (Exception ex) {
+      LOGGER.error("Could not read a resource: " + resourceName, ex);
+      return null;
+    }
+  }
+
+  @Nullable
+  private Properties getProperties(@Nullable InputStream stream) throws IOException {
+    if (stream == null) {
+      LOGGER.error("Stream is null.");
+      return null;
```

Comment:
I do not have a strong opinion on this, but sometimes returning `null` is considered a "not so good" practice. there are several approached on this: https://www.codebyamir.com/blog/stop-returning-null-in-java , however, I conceptually like `Exceptions` and maybe `Optional`s can be something to consider: https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

Reply:
I also prefer well-handled Exceptions over null-returns (and perhaps the most often heard argument against exceptions - that they "are slow" - is certainly NOT an issue here). I actually even started with a solution where Resources and BuildInfo throw exceptions which would then be handled by the activity class which can also print more meaningful logs.

But then there was a weird issue that the IDE (one I was running when testing) flashed a red **error** icon in the lower right corner of the window indicating that something was wrong when these exceptions were thrown (I tested with some invalid data) **even though** the exceptions were properly caught. (It sounds weird and I cannot easily reproduce this behavior so there's a possibility that i was mistaken, but I don't think so.)

Of course the IDE wasn't wrong: an exception was thrown because something was "wrong" (e.g. a resource file wasn't found). But I reasoned it is not the responsibility of this classes to decide if that is a big enough issue that it should be reported to the user. So after a moment of uncertainty about how to solve that I remembered a good ol' `null` that represents "nothing". 

Calls `getProperties("foo.properties")`: _Please, give me what you have on 'foo.properties'._

Returns `null`: _Sure, there is no such resource, so here is nothing._

Not sure if this is a good enough argument to keep null-returns but this what I reasoned when I wrote the code.

Also the `@NotNull` and `@Nullable` annotations helped a lot finding potential sources of `NullPointerException`s.